### PR TITLE
CLOUDSTACK-10056

### PR DESCRIPTION
**Pull request by nathanejohnson**
fix bug where disk controller specified via vm details throws
a NumberFormatException, since "scsi" is not a number.

### CODE REVIEW

File: server/src/com/cloud/vm/UserVmManagerImpl.java [hunk 1]
```diff
@@ -3619,8 +3619,12 @@ public UserVmVO doInTransaction(TransactionStatus status) throws InsufficientCap
 
                 _vmDao.persist(vm);
                 for (String key : customParameters.keySet()) {
-                    //handle double byte strings.
-                    vm.setDetail(key, Integer.toString(Integer.parseInt(customParameters.get(key))));
+                    if( key.equalsIgnoreCase(VmDetailConstants.ROOT_DISK_CONTROLLER) || key.equalsIgnoreCase(VmDetailConstants.DATA_DISK_CONTROLLER)) {
+                        vm.setDetail(key, customParameters.get(key));
+                    } else {
+                        // handle double byte strings.
+                        vm.setDetail(key, Integer.toString(Integer.parseInt(customParameters.get(key))));
```
Comment (rhtyd) [inline server/src/com/cloud/vm/UserVmManagerImpl.java]: In what cases would we do that @bvbharat (CLOUDSTACK-9638) ? do you think, the values set in database should be a string?

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
Alright, I compare the change with https://github.com/apache/cloudstack/pull/1967 LGTM

**Comment by nathanejohnson** [comment]
@rhtyd one other interesting piece, this same check does not seem to occur when registering a template.

**Comment by rhtyd** [comment]
@nathanejohnson @bvbharatk @karuturi I further debugged the issue, the original change from #1967 should be reverted or restricted to only parameters that are on integer type, by performing a deploy vm test, I think we should restrict the code that does string->int->string conversion to only these details: `cpuNumber`, `cpuSpeed` and `memory`, or better yet fix the issue in UI that is causing double-byte issue.

I deployed a VM with custom compute offering:
screenshot from 2017-08-25 15-40-32

And it sent the following params as details:
screenshot from 2017-08-25 15-41-18

**Comment by cloudmonger** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 1193
 Hypervisor xenserver
 NetworkType Advanced
 Passed=116
 Failed=5
 Skipped=40

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/r2si930m8xxzavs/AAAzNrnoF1fC3auFrvsKo_8-a?dl=0

**Failed tests:**
* test_scale_vm.py

 * ContextSuite context=TestScaleVm>:setup Failing since 47 runs

* test_routers_network_ops.py

 * test_01_isolate_network_FW_PF_default_routes_egress_true Failing since 53 runs

 * test_02_isolate_network_FW_PF_default_routes_egress_false Failing since 180 runs

 * test_01_RVR_Network_FW_PF_SSH_default_routes_egress_true Failing since 173 runs

 * test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false Failing since 173 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_01_verify_libvirt
test_02_verify_libvirt_after_restart
test_03_verify_libvirt_attach_disk
test_04_verify_guest_lspci
test_05_change_vm_ostype_restart
test_06_verify_guest_lspci_again
test_disable_oobm_ha_state_ineligible
test_hostha_configure_default_driver
test_hostha_enable_ha_when_host_disabled
test_hostha_enable_ha_when_host_disconected
test_hostha_enable_ha_when_host_in_maintenance
test_hostha_kvm_host_degraded
test_hostha_kvm_host_fencing
test_hostha_kvm_host_recovering
test_remove_ha_provider_not_possible
test_configure_ha_provider_invalid
test_configure_ha_provider_valid
test_ha_configure_enabledisable_across_clusterzones
test_ha_disable_feature_invalid
test_ha_enable_feature_invalid
test_ha_list_providers
test_ha_multiple_mgmt_server_ownership
test_ha_verify_fsm_available
test_ha_verify_fsm_degraded
test_ha_verify_fsm_fenced
test_ha_verify_fsm_recovering
test_hostha_configure_default_driver
test_hostha_configure_invalid_provider
test_hostha_disable_feature_valid
test_hostha_enable_feature_valid
test_hostha_enable_feature_without_setting_provider
test_list_ha_for_host
test_list_ha_for_host_invalid
test_list_ha_for_host_valid
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vm_snapshots.py
test_over_provisioning.py
test_global_settings.py
test_router_dnsservice.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_outofbandmanagement_nestedplugin.py
test_portforwardingrules.py
test_non_contigiousvlan.py
test_login.py
test_deploy_vm_iso.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_metrics_api.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_vm_life_cycle.py
test_disk_offerings.py

**Comment by rhtyd** [comment]
@bvbharat @bvbharatk comment please?
@nathanejohnson can you change the implementation to restrict the double bye handling code to work only for details keys - `cpuNumber`, `cpuSpeed` and `memory`? For rest we should simply read the string value with no addtional transformations.

**Comment by nathanejohnson** [comment]
@rthyd that sounds like a better approach.  I'll get this updated ASAP

File: server/src/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -3619,8 +3619,14 @@ public UserVmVO doInTransaction(TransactionStatus status) throws InsufficientCap
...
                 _vmDao.persist(vm);
                 for (String key : customParameters.keySet()) {
-                    //handle double byte strings.
-                    vm.setDetail(key, Integer.toString(Integer.parseInt(customParameters.get(key))));
+                    if( key.equalsIgnoreCase(VmDetailConstants.CPU_NUMBER) ||
+                            key.equalsIgnoreCase(VmDetailConstants.CPU_SPEED) ||
+                            key.equalsIgnoreCase(VmDetailConstants.MEMORY)) {
+                        // handle double byte strings.
+                        vm.setDetail(key, Integer.toString(Integer.parseInt(customParameters.get(key))));
+                    } else {
```
Comment (rhtyd) [inline server/src/com/cloud/vm/UserVmManagerImpl.java]: Fantastic thanks for fixing @nathanejohnson

**Comment by rhtyd** [comment]
LGTM, I'll kick one final test run and then we can merge based on test results. Thanks @nathanejohnson 
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by GabrielBrascher** [comment]
Thanks @nathanejohnson. Code LGTM!

**Comment by rhtyd** [comment]
Test LGTM, merging now. Thanks @nathanejohnson

**Pull request merged**